oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/optimizer/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,151 @@
+package com.empowerops.volition.ref_oasis.optimizer
+
+import com.empowerops.volition.dto.*
+import com.empowerops.volition.ref_oasis.model.ModelService
+import com.empowerops.volition.ref_oasis.model.hasName
+import io.grpc.Status
+import io.grpc.StatusRuntimeException
+import io.grpc.stub.StreamObserver
+
+class OptimizerEndpoint(
+        private val apiService: IApiService,
+        private val modelService: ModelService
+) : OptimizerGrpc.OptimizerImplBase() {
+
+    override fun changeNodeName(
+            request: NodeNameChangeCommandDTO,
+            responseObserver: StreamObserver<NodeNameChangeResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.oldName)) {
+            apiService.changeNodeName(request)
+        }
+    }
+
+    override fun autoConfigure(
+            request: NodeStatusCommandOrResponseDTO,
+            responseObserver: StreamObserver<NodeChangeConfirmDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.autoConfigure(request)
+        }
+    }
+
+    override fun updateConfiguration(
+            request: ConfigurationCommandDTO,
+            responseObserver: StreamObserver<ConfigurationResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.updateConfiguration(request)
+        }
+    }
+
+    override fun requestRunResult(
+            request: ResultRequestDTO,
+            responseObserver: StreamObserver<ResultResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.resultRequest(request)
+        }
+    }
+
+    override fun registerRequest(
+            request: RequestRegistrationCommandDTO,
+            responseObserver: StreamObserver<RequestQueryDTO>
+    ) {
+            apiService.register(request, responseObserver)
+    }
+
+    override fun startOptimization(
+            request: StartOptimizationCommandDTO,
+            responseObserver: StreamObserver<StartOptimizationResponseDTO>
+    ) = responseObserver.consumeThen({
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.requestStart(request)
+        }
+    }) { response ->
+        if (response.acknowledged) apiService.start()
+    }
```

Comment:
yeah this is pretty funny. The RX-java guys get into this kind of thing a lot. I dont like it.

I think the key thing you're trying to express here is that you want to be able to run some code to generate the result, and then run some code after its sent. 

this issue of "after it is sent" is actually a rather difficult one, since it might be hard for a test to properly synchronize ont hat, but if you're mocking the endpoint it should be feasible. Of course, in production, the distinction is irrelevant since after you call `onNext()` GRPC makes no gaurentees about when that message will be delivered, and, of course, you have no way of knowing when the other side will actually finish reading the result.

But, the key point here is that we have 2 stages. I propose we use a channel to match these two stages. Between the first stge and the second stage we'll emit a value on the channel, and to signal the end of the second stage we'll close the channel.

thus the usage I want is

```kt
override fun startOptimization(
  val request: StartOptimizationRequestDTO,
): RecieveChannel<StartOptimizationResponseDTO> {
  return produce<StartOptimizationResponseDTO>{
    val response = checkThenRun(hasName()) { requestStart() }
    send(response)
    if(response.acknowledged) apiService.start()
  }
}
```

similarly, reading that code its kind've strange. I suspect this is because you needed to write the second lambda with only the context of the DTO you returned from the first lambda. Since we no longer have that requirement, we can re-write this a bit more naturally IMHO:

```kt
override fun startOptimization(
  val request: StartOptimizationRequestDTO,
): RecieveChannel<StartOptimizationResponseDTO> {
  return produce<StartOptimizationResponseDTO>{
    if(hasName()){ 
      val response = StartOptimizationResponseDTO().apply { acknoledged = false }
      send(response)
    }
    else {
      val response = requestStart()
      send(response)
      apiService.start()
    }
  }
}
```

and finally we get the implementations

```kt

//extensions.kt
fun <T> StreamObserver<T>.consumeSingleFrom(producer: ReceiveChannel<T>) {
  //globalscope.launch...
    try {
        val response = producer.recieveOrNull()
        onNext(result ?: throw IllegalStateException("producer closed before yielding a response"))
        val secondResponse = producer.recieveOrNull() 
        if(secondResponse != null) throw IlegalStateException("produced second response: $secondResponse")
        require(producer.isClosedForReceive) //this is redundant but I wanna make sure the channel is dead as a door-nail. 
        onCompleted()
    } catch (ex: Exception) {
        onError(ex)
        throw ex
    } finally {
        onCompleted()
    }
}

//optimizerEndpoint

override fun startOptimization(request: RequestDTO, response: ResponseObserver<ResponseDTO>) 
    = response.consumeSingleFrom(apiService.startOptimization(request));
```